dali/imgcodec/decoders/tiff_libtiff.cc

```diff
@@ -0,0 +1,228 @@
+  LIBTIFF_CALL(TIFFGetFieldDefaulted(tiffptr, TIFFTAG_BITSPERSAMPLE, &info.bit_depth));
+  LIBTIFF_CALL(TIFFGetFieldDefaulted(tiffptr, TIFFTAG_ORIENTATION, &info.orientation));
+  LIBTIFF_CALL(TIFFGetFieldDefaulted(tiffptr, TIFFTAG_COMPRESSION, &info.compression));
+  LIBTIFF_CALL(TIFFGetFieldDefaulted(tiffptr, TIFFTAG_ROWSPERSTRIP, &info.rows_per_strip));
+
+  info.is_tiled = TIFFIsTiled(tiffptr);
+
+  uint16_t photometric_interpretation;
+  if (TIFFGetField(tiffptr, TIFFTAG_PHOTOMETRIC, &photometric_interpretation) &&
+      photometric_interpretation == PHOTOMETRIC_PALETTE) {
+    info.is_palette = true;
+  }
+
+  return info;
+}
+
+}  // namespace detail
+
+DecodeResult LibTiffDecoderInstance::Decode(SampleView<CPUBackend> out, ImageSource *in,
+                                           DecodeParams opts, const ROI &requested_roi) {
+  auto tiff = detail::OpenTiff(in);
+  auto info = detail::GetTiffInfo(tiff.get());
+
+  // TODO(skarpinski) Support other bit depths
+  DALI_ENFORCE(info.bit_depth == 8, "Not implemented: Only bit depth = 8 is supported");
+
+  // TODO(skarpinski) Support palette images
+  DALI_ENFORCE(!info.is_palette, "Not implemented: Palette images are not yet supported");
+
+  // TODO(skarpinski) Support tiled images
+  DALI_ENFORCE(!info.is_tiled, "Not implemented: Tiled images are not yet supported");
+
+  unsigned out_channels = NumberOfChannels(opts.format, info.channels);
+
+  // TODO(skarpinski) support different types
+  using InType = uint8_t;
+  using OutType = uint8_t;
+
+  auto row_nbytes = TIFFScanlineSize(tiff.get());
+  std::unique_ptr<InType, void(*)(void*)> row_buf{
+    static_cast<InType *>(_TIFFmalloc(row_nbytes)), _TIFFfree};
+  DALI_ENFORCE(row_buf.get() != nullptr, "Could not allocate memory");
+  InType * const row_in  = row_buf.get();
+  OutType * const img_out = out.mutable_data<OutType>();
+
+  ROI roi;
+  if (!requested_roi.use_roi()) {
+    roi.begin = {0, 0};
+    roi.end = out.shape().first(2);
+  } else {
+    roi = requested_roi;
+  }
+
+  // Need to read sequentially since not all the images support random access
+  // From: http://www.libtiff.org/man/TIFFReadScanline.3t.html
+  // Compression algorithm does not support random access. Data was requested in a non-sequential
+  // order from a file that uses a compression algorithm and that has RowsPerStrip greater than one.
+  // That is, data in the image is stored in a compressed form, and with multiple rows packed into a
+  // strip. In this case, the library does not support random access to the data. The data should
+  // either be accessed sequentially, or the file should be converted so that each strip is made up
+  // of one row of data.
+  const bool allow_random_row_access = (info.compression == COMPRESSION_NONE
+                                        || info.rows_per_strip == 1);
+  if (!allow_random_row_access) {
+    for (int64_t y = 0; y < roi.begin[0]; y++) {
```

Comment:
I liked the comment from the original one, that explained that we are skipping the rows because of the mandated sequential access.

Reply:
Added the comment